pkg/proxy/proxy.go

```diff
@@ -412,6 +412,7 @@ func (p *Proxy) CreateOrUpdateRedirect(l4 policy.ProxyPolicy, id string, localEn
 			var implUpdateRevertFunc revert.RevertFunc
 			implUpdateRevertFunc, err = redir.implementation.UpdateRules(wg)
 			if err != nil {
+				redir.mutex.Unlock()
```

Comment:
Wouldn't it be possible to use a `defer` call here?

Reply:
There already is a `defer ` call which includes `p.mutex.Unlock()` further up at line 395 (cannot put comment there 🤔), so I think adding this here would lead to a double unlock.

Reply:
This seems really odd indeed. Be aware this is not `p.mutex.Unlock()` but is `redir.mutex.Unlock()`.

Reply:
Oops, good point. Got confused by the nested locking.

Reply:
To answer to Paul, I don't think we could use a `defer` because the function returned in `updateRevertFunc := redir.updateRules(l4)` needs the mutex, so we want to release it before we carry on in case we have to call the `revertFunc()`.

Reply:
Ok, fair enough. I guess we can always change that if we later break this function down.